jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -9,825 +9,181 @@
+    non-witness input fields are roughly 32+4+4+20+4=64, so total becomes
+    n_in * 64 + 4(ver) + 4(locktime) + n_out*34
+    Assuming p2wpkh native:
+    witness as previous case
+    non-witness loses the 24 witnessprogram, replaced with 1 zero,
+    in the scriptSig, so becomes:
+    n_in * 41 + 4(ver) + 4(locktime) +2 (len in, out) + n_out*34
+    '''
+    if txtype == 'p2pkh':
+        return 10 + ins * 147 + 34 * outs
+    elif txtype == 'p2sh-p2wpkh':
+        #return the estimate for the witness and non-witness
+        #portions of the transaction, assuming that all the inputs
+        #are of segwit type p2sh-p2wpkh
+        # Note as of Jan19: this misses 2 bytes (trivial) for len in, out
+        # and also overestimates output size by 2 bytes.
+        witness_estimate = ins*109
+        non_witness_estimate = 4 + 4 + outs*34 + ins*64
+        return (witness_estimate, non_witness_estimate)
+    elif txtype == 'p2wpkh':
+        witness_estimate = ins*109
+        non_witness_estimate = 4 + 4 + 2 + outs*31 + ins*41
+        return (witness_estimate, non_witness_estimate)
+    elif txtype == 'p2shMofN':
+        ins, M, N = ins
+        return 10 + (45 + 74*M + 34*N) * ins + 34 * outs
+    else:
+        raise NotImplementedError("Transaction size estimation not" +
+                                  "yet implemented for type: " + txtype)
 
 def pubkey_to_p2pkh_script(pub, require_compressed=False):
-    """ Construct a pay-to-pubkey-hash scriptPubKey
-    given a single pubkey. Script returned in binary.
-    The require compressed flag may be used to disallow
-    uncompressed keys, e.g. in constructing a segwit
-    scriptCode field.
     """
-    if not isinstance(pub, bytes):
-        pub = binascii.unhexlify(pub)
-    return pubkey_to_script(pub, P2PKH_PRE, P2PKH_POST)
-
-def pubkey_to_p2sh_p2wpkh_script(pub):
-    """ Construct a nested-pay-to-witness-pubkey-hash
-    scriptPubKey given a single pubkey.
-    Script returned in binary.
+    Given a pubkey in bytes, return a CScript
+    representing the corresponding pay-to-pubkey-hash
+    scriptPubKey.
     """
-    if not isinstance(pub, bytes):
-        pub = binascii.unhexlify(pub)
-    wscript = pubkey_to_p2wpkh_script(pub)
-    return P2SH_P2WPKH_PRE + bin_hash160(wscript) + P2SH_P2WPKH_POST
-
-def pubkey_to_p2sh_p2wpkh_address(pub, magicbyte=5):
-    """ Construct a nested-pay-to-witness-pubkey-hash
-    address given a single pubkey; magicbyte defines
-    network as for any p2sh address.
-    """    
-    if not isinstance(pub, bytes):
-        pub = binascii.unhexlify(pub)
-    script = pubkey_to_p2wpkh_script(pub)
-    return p2sh_scriptaddr(script, magicbyte=magicbyte)
+    if not is_valid_pubkey(pub, require_compressed=require_compressed):
+        raise Exception("Invalid pubkey")
+    return CScript([OP_DUP, OP_HASH160, Hash160(pub),
+                                          OP_EQUALVERIFY, OP_CHECKSIG])
```

Comment:
Can also be done with `P2PKHCoinAddress.from_pubkey(pub).to_scriptPubKey()`
Alternatively, can be done with `bitcointx.core.script.standard_keyhash_scriptpubkey(Hash160(pub))`

Reply:
Thanks, this is a very useful suggestion, it is much better if I defer the lower level construction to the underlying lib. Addressed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/1e2fc3848295a05f4ee685fd435f277c378acadc